runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/DoFnInfo.java

```diff
@@ -17,44 +17,121 @@
   private final Iterable<PCollectionView<?>> sideInputViews;
   private final Coder<InputT> inputCoder;
   private final long mainOutput;
   private final Map<Long, TupleTag<?>> outputMap;
 
-  public DoFnInfo(OldDoFn<InputT, OutputT> doFn,
+  /** Lazily initialized invoker for this {@link DoFn}. */
+  @Nullable
+  private transient DoFnInvoker<InputT, OutputT> doFnInvoker;
+
+  /** Creates a {@link DoFnInfo} for the given {@link DoFn} and auxiliary bits and pieces. */
+  public DoFnInfo(
+      DoFn<InputT, OutputT> doFn,
+      WindowingStrategy<?, ?> windowingStrategy,
+      Iterable<PCollectionView<?>> sideInputViews,
+      Coder<InputT> inputCoder,
+      long mainOutput,
+      Map<Long, TupleTag<?>> outputMap) {
+    this((Serializable) doFn, windowingStrategy, sideInputViews, inputCoder, mainOutput, outputMap);
+  }
+
+  /** Uses of this constructor should be ported to pass a {@link DoFn}. */
+  @Deprecated
+  public DoFnInfo(
+      OldDoFn<InputT, OutputT> oldDoFn,
+      WindowingStrategy<?, ?> windowingStrategy,
+      Iterable<PCollectionView<?>> sideInputViews,
+      Coder<InputT> inputCoder,
+      long mainOutput,
+      Map<Long, TupleTag<?>> outputMap) {
+    this(
+        (Serializable) oldDoFn,
+        windowingStrategy,
+        sideInputViews,
+        inputCoder,
+        mainOutput,
+        outputMap);
+  }
+
+  /** This constructor provides a migration path from {@link OldDoFn} to {@link DoFn}. */
+  @Deprecated
+  public DoFnInfo(
+      Serializable oldDoFnOrDoFn,
       WindowingStrategy<?, ?> windowingStrategy,
       Iterable<PCollectionView<?>> sideInputViews,
       Coder<InputT> inputCoder,
       long mainOutput,
       Map<Long, TupleTag<?>> outputMap) {
-    this.doFn = doFn;
+    checkArgument(
+        oldDoFnOrDoFn instanceof DoFn || oldDoFnOrDoFn instanceof OldDoFn,
+        "%s fn argument must be a %s or an %s",
+        getClass().getSimpleName(),
+        DoFn.class.getSimpleName(),
+        OldDoFn.class.getSimpleName());
+    this.oldDoFnOrDoFn = oldDoFnOrDoFn;
     this.windowingStrategy = windowingStrategy;
     this.sideInputViews = sideInputViews;
     this.inputCoder = inputCoder;
     this.mainOutput = mainOutput;
     this.outputMap = outputMap;
   }
 
+  /**
+   * @deprecated callers should use {@link #getDoFnInvoker()} if possible, to be agnostic as to
+   *     whether this contains a {@link DoFn} or an {@link OldDoFn}. If access to the actual object
+   *     is needed, then temporarily use {@link #getOldDoFnOrDoFn()} until this method is ported to
+   *     return {@link DoFn} instead of {@link OldDoFn}.
+   */
+  @Deprecated
   public OldDoFn<InputT, OutputT> getDoFn() {
-    return doFn;
+    checkState(oldDoFnOrDoFn instanceof OldDoFn,
+        "%s.getDoFn() called when it did not contain %s",
+        DoFnInfo.class.getSimpleName(), OldDoFn.class.getSimpleName());
+    return (OldDoFn) oldDoFnOrDoFn;
+  }
+
+  public DoFnInvoker<InputT, OutputT> getDoFnInvoker() {
```

Comment:
Since we want to have DoFnInfo just a serialized blob with no business logic, wouldn't it make sense to update the Dataflow worker to cast to call this method expecting its getting a Serializable instead of an OldDoFn and call DoFnInvokers.INSTANCE.invokerFor(Serializable)?

Then this update could just change the field type to be Serializable for DoFnInfo and the return type for getDoFn to return Serializable.

This would allow you to build a new worker image associated with this change as well that is Serializable aware for the getDoFn method upfront.